// @bostongene/frontend-components/Textarea/index.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var react_1 = __importStar(require("react"));
var classnames_1 = __importDefault(require("classnames"));
var isEqual_1 = __importDefault(require("lodash/isEqual"));
require("./Textarea.scss");
var Textarea = (function (_super) {
    __extends(Textarea, _super);
    function Textarea() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.state = { value: _this.props.value };
        _this.textarea = react_1.createRef();
        _this.update = function (value) { return _this.setState({ value: value }); };
        _this.onChange = function (_a) {
            var target = _a.target;
            var onChange = _this.props.onChange;
            _this.setState({ value: target.value }, function () { return onChange(target); });
        };
        return _this;
    }
    Textarea.prototype.componentDidUpdate = function (prevProps) {
        var _a;
        var value = this.props.value;
        if ((_a = this.textarea.current) === null || _a === void 0 ? void 0 : _a.contains(document.activeElement))
            return;
        if (!isEqual_1.default(value, prevProps.value))
            this.update(value);
    };
    Textarea.prototype.render = function () {
        var _a;
        var _b = this.props, id = _b.id, name = _b.name, className = _b.className, placeholder = _b.placeholder, autoComplete = _b.autoComplete, disabled = _b.disabled, hasError = _b.hasError;
        var value = this.state.value;
        var textareaWrapperClasses = classnames_1.default((_a = {
                'bg-textarea-wrapper': true,
                'bg-textarea-wrapper--error': hasError
            },
            _a[className] = className,
            _a));
        return (react_1.default.createElement("div", { className: textareaWrapperClasses },
            react_1.default.createElement("textarea", { ref: this.textarea, id: id, name: name, value: value, className: "bg-textarea", autoComplete: autoComplete, placeholder: placeholder, disabled: disabled, onChange: this.onChange })));
    };
    Textarea.defaultProps = {
        value: '',
        className: '',
        placeholder: '',
        autoComplete: 'off',
        disabled: false,
        hasError: false,
        onChange: function () { }
    };
    return Textarea;
}(react_1.PureComponent));
exports.default = Textarea;
