// @bostongene/frontend-components/NumberInput/index.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
var react_1 = __importStar(require("react"));
var classname_1 = require("@bem-react/classname");
require("./NumberInput.scss");
var inputWrapperClassName = classname_1.withNaming({ e: '__', m: '--' })('bg-number-input-wrapper');
var NumberInput = (function (_super) {
    __extends(NumberInput, _super);
    function NumberInput() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.state = { value: "" + _this.props.value };
        _this.input = react_1.createRef();
        _this.update = function (value) { return _this.setState({ value: value }); };
        _this.onChange = function (target) {
            var onChange = _this.props.onChange;
            _this.setState({ value: target.value }, function () { return onChange(target); });
        };
        return _this;
    }
    NumberInput.prototype.componentDidUpdate = function (prevProps) {
        var _a;
        var value = this.props.value;
        if ((_a = this.input.current) === null || _a === void 0 ? void 0 : _a.contains(document.activeElement))
            return;
        if ("" + value !== prevProps.value)
            this.update("" + value);
    };
    NumberInput.prototype.render = function () {
        var _this = this;
        var _a = this.props, id = _a.id, name = _a.name, className = _a.className, placeholder = _a.placeholder, autoComplete = _a.autoComplete, maxValue = _a.maxValue, minValue = _a.minValue, step = _a.step, disabled = _a.disabled, hasError = _a.hasError;
        var value = this.state.value;
        var inputWrapperClasses = inputWrapperClassName('', { error: hasError }, [className]);
        return (react_1.default.createElement("div", { className: inputWrapperClasses },
            react_1.default.createElement("input", { ref: this.input, type: "number", id: id, name: name, value: value, className: "bg-number-input", autoComplete: autoComplete, placeholder: placeholder, max: maxValue, min: minValue, step: step, disabled: disabled, onChange: function (_a) {
                    var target = _a.target;
                    return _this.onChange(target);
                } })));
    };
    NumberInput.defaultProps = {
        value: '',
        className: '',
        placeholder: '',
        autoComplete: 'off',
        disabled: false,
        hasError: false,
        onChange: function () { }
    };
    return NumberInput;
}(react_1.PureComponent));
exports.default = NumberInput;
