// @bostongene/frontend-components/Input/index.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var react_1 = __importStar(require("react"));
var classnames_1 = __importDefault(require("classnames"));
require("./Input.scss");
var crossIcon = function () { return (react_1.default.createElement("svg", { xmlns: "http://www.w3.org/2000/svg", width: "18", height: "18", viewBox: "0 0 18 18" },
    react_1.default.createElement("g", { fill: "none", fillRule: "evenodd", transform: "translate(-3 -3)" },
        react_1.default.createElement("circle", { cx: "12", cy: "12", r: "9", fillOpacity: ".6" }),
        react_1.default.createElement("path", { d: "M11.97 10.556l1.98-1.98a.85.85 0 1 1 1.202 1.202l-1.98 1.98 1.98 1.98a.85.85 0 1 1-1.202 1.202l-1.98-1.98-1.98 1.98a.85.85 0 1 1-1.202-1.202l1.98-1.98-1.98-1.98A.85.85 0 1 1 9.99 8.576l1.98 1.98z" })))); };
var Input = (function (_super) {
    __extends(Input, _super);
    function Input() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.state = { value: _this.props.value };
        _this.input = react_1.createRef();
        _this.update = function (value) { return _this.setState({ value: value }); };
        _this.onChange = function (target) {
            var onChange = _this.props.onChange;
            _this.setState({ value: target.value }, function () { return onChange(target); });
        };
        _this.onReset = function () {
            var _a = _this.props, onChange = _a.onChange, name = _a.name;
            _this.setState({ value: '' }, function () {
                var _a;
                onChange({ name: name, value: '' });
                (_a = _this.input.current) === null || _a === void 0 ? void 0 : _a.focus();
            });
        };
        return _this;
    }
    Input.prototype.componentDidUpdate = function (prevProps) {
        var _a;
        var value = this.props.value;
        if ((_a = this.input.current) === null || _a === void 0 ? void 0 : _a.contains(document.activeElement))
            return;
        if (value !== prevProps.value)
            this.update(value);
    };
    Input.prototype.render = function () {
        var _a;
        var _this = this;
        var _b = this.props, id = _b.id, name = _b.name, className = _b.className, placeholder = _b.placeholder, autoComplete = _b.autoComplete, disabled = _b.disabled, clearable = _b.clearable, hasError = _b.hasError;
        var value = this.state.value;
        var inputWrapperClasses = classnames_1.default((_a = {
                'bg-input-wrapper': true,
                'bg-input-wrapper--error': hasError
            },
            _a[className] = className,
            _a));
        var inputClasses = classnames_1.default({
            'bg-input': true,
            'bg-input--clearable': clearable
        });
        return (react_1.default.createElement("div", { className: inputWrapperClasses },
            react_1.default.createElement("input", { ref: this.input, id: id, name: name, value: value, className: inputClasses, autoComplete: autoComplete, placeholder: placeholder, disabled: disabled, type: "text", onChange: function (_a) {
                    var target = _a.target;
                    return _this.onChange(target);
                } }),
            clearable && value && (react_1.default.createElement("div", { className: "bg-input-wrapper__cross-icon", onClick: this.onReset }, crossIcon()))));
    };
    Input.defaultProps = {
        value: '',
        className: '',
        placeholder: '',
        autoComplete: 'off',
        disabled: false,
        clearable: false,
        hasError: false,
        onChange: function () { }
    };
    return Input;
}(react_1.PureComponent));
exports.default = Input;
